services/src/contexts/mod.rs

```diff
@@ -37,7 +57,132 @@ pub trait Context: 'static + Send + Sync + Clone {
+}
+
+pub struct QueryContextImpl {
+    chunk_byte_size: usize,
+}
+
+impl QueryContext for QueryContextImpl {
+    fn chunk_byte_size(&self) -> usize {
+        self.chunk_byte_size
+    }
+}
+
+pub struct ExecutionContextImpl<D>
+where
+    D: DataSetDB,
+{
+    data_set_db: DB<D>,
+    thread_pool: Arc<ThreadPool>,
+    user: UserId,
+}
+
+impl<D> ExecutionContext for ExecutionContextImpl<D>
+where
+    D: DataSetDB
+        + MetaDataProvider<MockDataSetDataSourceLoadingInfo, VectorResultDescriptor>
+        + MetaDataProvider<OgrSourceDataset, VectorResultDescriptor>,
+{
+    fn thread_pool(&self) -> ThreadPoolContext {
+        self.thread_pool.create_context()
+    }
+
+    fn raster_data_root(&self) -> std::result::Result<PathBuf, geoengine_operators::error::Error> {
+        Ok(get_config_element::<GdalSource>()
+            .map_err(|_| geoengine_operators::error::Error::RasterRootPathNotConfigured)?
+            .raster_data_root_path)
+    }
+
+    fn tiling_specification(&self) -> TilingSpecification {
+        // TODO: load only once and handle error
+        let config_tiling_spec = get_config_element::<config::TilingSpecification>().unwrap();
+
+        TilingSpecification {
+            origin_coordinate: Coordinate2D::new(
+                config_tiling_spec.origin_coordinate_x,
+                config_tiling_spec.origin_coordinate_y,
+            ),
+            tile_size_in_pixels: GridShape2D::from([
+                config_tiling_spec.tile_shape_pixels_y,
+                config_tiling_spec.tile_shape_pixels_x,
+            ]),
+        }
+    }
+}
+
+// TODO: use macro(?) for delegating meta_data function to DataSetDB to avoid redundant code
+impl<D> MetaDataProvider<MockDataSetDataSourceLoadingInfo, VectorResultDescriptor>
+    for ExecutionContextImpl<D>
+where
+    D: DataSetDB + MetaDataProvider<MockDataSetDataSourceLoadingInfo, VectorResultDescriptor>,
+{
+    // TODO: make async
+    fn meta_data(
+        &self,
+        data_set: &DataSetId,
+    ) -> Result<
+        Box<dyn MetaData<MockDataSetDataSourceLoadingInfo, VectorResultDescriptor>>,
+        geoengine_operators::error::Error,
+    > {
+        futures::executor::block_on(async {
+            match data_set {
+                DataSetId::Internal(_) => self.data_set_db.read().await.meta_data(data_set),
+                DataSetId::Staging(_) => todo!(),
+                DataSetId::External(external) => self
+                    .data_set_db
+                    .read()
+                    .await
+                    .data_set_provider(self.user, external.provider)
+                    .await
+                    .map_err(|e| geoengine_operators::error::Error::LoadingInfo {
+                        source: Box::new(e),
+                    })?
+                    .meta_data(data_set),
+            }
+        })
+    }
+}
+
+// TODO: use macro(?) for delegating meta_data function to DataSetDB to avoid redundant code
```

Comment:
I think it would just be a proxy. I tried to implement it using generics but there are no appropriate bounds to use for generic implementation because the DataSetProvider only implements very specific generics which have nothing in common. Not sure if auto-impl would work.